File: tests/spike/graphs.js
/**
 * Spike Test Graph Generator
 * Parses JTL files and generates performance graphs:
 * - Response time percentiles (50th, 95th)
 * - Error rate over time
 * - Active threads over time
 * - Throughput
 */

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const RESULTS_DIR = path.join(__dirname, "results");

/**
 * Parse JTL CSV file
 */
const parseJTL = (filePath) => {
  if (!fs.existsSync(filePath)) return null;
  
  const data = fs.readFileSync(filePath, "utf-8");
  const lines = data.trim().split("\n");
  
  if (lines.length < 2) return null;
  
  const header = lines[0].split(",");
  const timeStampIdx = header.indexOf("timeStamp");
  const elapsedIdx = header.indexOf("elapsed");
  const successIdx = header.indexOf("success");
  const grpThreadsIdx = header.indexOf("grpThreads");
  
  const samples = [];
  for (let i = 1; i < lines.length; i++) {
    const parts = lines[i].split(",");
    if (parts.length < elapsedIdx + 1) continue;
    
    samples.push({
      timestamp: parseInt(parts[timeStampIdx]) || 0,
      elapsed: parseInt(parts[elapsedIdx]) || 0,
      success: parts[successIdx]?.toLowerCase() === "true",
      activeThreads: parseInt(parts[grpThreadsIdx]) || 0,
    });
  }
  
  return samples;
};

/**
 * Calculate percentile
 */
const percentile = (arr, p) => {
  if (arr.length === 0) return 0;
  const sorted = arr.sort((a, b) => a - b);
  const index = Math.ceil((p / 100) * sorted.length) - 1;
  return sorted[Math.max(0, index)];
};

/**
 * Generate time-series data binned by time intervals
 */
const generateTimeSeries = (samples, binIntervalMs = 1000) => {
  if (samples.length === 0) return [];
  
  const minTime = Math.min(...samples.map(s => s.timestamp));
  const maxTime = Math.max(...samples.map(s => s.timestamp));
  const bins = [];
  
  for (let t = minTime; t <= maxTime; t += binIntervalMs) {
    const binEnd = t + binIntervalMs;
    const binSamples = samples.filter(s => s.timestamp >= t && s.timestamp < binEnd);
    
    if (binSamples.length === 0) continue;
    
    const responseTimes = binSamples.map(s => s.elapsed);
    const successCount = binSamples.filter(s => s.success).length;
    const errorCount = binSamples.length - successCount;
    
    bins.push({
      time: new Date(t).toISOString(),
      timestamp: t,
      p50: percentile(responseTimes, 50),
      p95: percentile(responseTimes, 95),
      avgResponse: Math.round(responseTimes.reduce((a, b) => a + b, 0) / responseTimes.length),
      errorRate: (errorCount / binSamples.length * 100).toFixed(2),
      throughput: binSamples.length,
      activeThreads: binSamples[0]?.activeThreads || 0,
      sampleCount: binSamples.length,
    });
  }
  
  return bins;
};

/**
 * Generate ASCII graph
 */
const generateASCIIGraph = (title, data, valueKey, maxWidth = 80, maxHeight = 20) => {
  console.log(`\n╔${"═".repeat(maxWidth - 2)}╗`);
  console.log(`║ ${title.padEnd(maxWidth - 4)} ║`);
  console.log(`╠${"═".repeat(maxWidth - 2)}╣`);
  
  if (data.length === 0) {
    console.log(`║ No data available${" ".repeat(maxWidth - 20)} ║`);
    console.log(`╚${"═".repeat(maxWidth - 2)}╝`);
    return;
  }
  
  const values = data.map(d => parseFloat(d[valueKey]) || 0);
  const maxValue = Math.max(...values, 1);
  const scale = maxHeight / maxValue;
  
  // Draw graph from top to bottom
  for (let h = maxHeight; h > 0; h--) {
    let line = "║ ";
    for (let i = 0; i < data.length; i++) {
      const barHeight = Math.round(values[i] * scale);
      line += barHeight >= h ? "█" : " ";
    }
    line += " ║";
    console.log(line);
  }
  
  console.log(`╠${"═".repeat(maxWidth - 2)}╣`);
  console.log(`║ Max: ${maxValue.toFixed(0).padEnd(10)} | Min: ${Math.min(...values).toFixed(0).padEnd(10)}${" ".repeat(maxWidth - 50)}║`);
  console.log(`╚${"═".repeat(maxWidth - 2)}╝`);
};

/**
 * Generate HTML chart
 */
const generateHTMLChart = (testName, timeSeries) => {
  const p50Values = timeSeries.map(t => t.p50);
  const p95Values = timeSeries.map(t => t.p95);
  const errorRates = timeSeries.map(t => parseFloat(t.errorRate));
  const activeThreads = timeSeries.map(t => t.activeThreads);
  const times = timeSeries.map((t, i) => i);
  
  const maxResponse = Math.max(...p50Values, ...p95Values);
  const maxThreads = Math.max(...activeThreads, 1);
  
  const html = `<!DOCTYPE html>
<html>
<head>
  <title>Spike Test: ${testName}</title>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/3.9.1/chart.min.js"></script>
  <style>
    body { font-family: Arial, sans-serif; margin: 20px; background: #f5f5f5; }
    .container { max-width: 1200px; margin: 0 auto; }
    .chart-section { background: white; padding: 20px; margin: 20px 0; border-radius: 8px; box-shadow: 0 2px 4px rgba(0,0,0,0.1); }
    h1 { color: #333; text-align: center; }
    .stats { display: grid; grid-template-columns: repeat(4, 1fr); gap: 15px; margin: 20px 0; }
    .stat-box { background: #f0f0f0; padding: 15px; border-radius: 5px; text-align: center; }
    .stat-value { font-size: 24px; font-weight: bold; color: #2196F3; }
    .stat-label { font-size: 12px; color: #666; margin-top: 5px; }
  </style>
</head>
<body>
  <div class="container">
    <h1>📊 Spike Test: ${testName}</h1>
    
    <div class="stats">
      <div class="stat-box">
        <div class="stat-value">${Math.max(...p50Values).toFixed(0)}ms</div>
        <div class="stat-label">Max P50 Response</div>
      </div>
      <div class="stat-box">
        <div class="stat-value">${Math.max(...p95Values).toFixed(0)}ms</div>
        <div class="stat-label">Max P95 Response</div>
      </div>
      <div class="stat-box">
        <div class="stat-value">${Math.max(...errorRates).toFixed(1)}%</div>
        <div class="stat-label">Max Error Rate</div>
      </div>
      <div class="stat-box">
        <div class="stat-value">${Math.max(...activeThreads)}</div>
        <div class="stat-label">Peak Threads</div>
      </div>
    </div>

    <div class="chart-section">
      <h2>Response Time (P50 & P95)</h2>
      <canvas id="responseChart"></canvas>
    </div>

    <div class="chart-section">
      <h2>Error Rate Over Time</h2>
      <canvas id="errorChart"></canvas>
    </div>

    <div class="chart-section">
      <h2>Active Threads Over Time</h2>
      <canvas id="threadsChart"></canvas>
    </div>

    <div class="chart-section">
      <h2>Throughput (Requests/Second)</h2>
      <canvas id="throughputChart"></canvas>
    </div>
  </div>

  <script>
    const p50Data = ${JSON.stringify(p50Values)};
    const p95Data = ${JSON.stringify(p95Values)};
    const errorData = ${JSON.stringify(errorRates)};
    const threadData = ${JSON.stringify(activeThreads)};
    const throughputData = ${JSON.stringify(timeSeries.map(t => t.throughput))};
    const timeLabels = ${JSON.stringify(times.slice(0, Math.min(timeSeries.length, 50)))};

    // Sample every Nth point for readability
    const sampleRate = Math.max(1, Math.floor(timeSeries.length / 50));
    const sampledP50 = p50Data.filter((_, i) => i % sampleRate === 0);
    const sampledP95 = p95Data.filter((_, i) => i % sampleRate === 0);
    const sampledErrors = errorData.filter((_, i) => i % sampleRate === 0);
    const sampledThreads = threadData.filter((_, i) => i % sampleRate === 0);
    const sampledThroughput = throughputData.filter((_, i) => i % sampleRate === 0);

    // Response Time Chart
    new Chart(document.getElementById('responseChart'), {
      type: 'line',
      data: {
        labels: timeLabels.slice(0, sampledP50.length),
        datasets: [
          {
            label: 'P50 Response Time',
            data: sampledP50,
            borderColor: '#2196F3',
            backgroundColor: 'rgba(33, 150, 243, 0.1)',
            tension: 0.3,
          },
          {
            label: 'P95 Response Time',
            data: sampledP95,
            borderColor: '#FF9800',
            backgroundColor: 'rgba(255, 152, 0, 0.1)',
            tension: 0.3,
          },
        ],
      },
      options: {
        responsive: true,
        plugins: { legend: { position: 'top' } },
        scales: { y: { title: { display: true, text: 'Response Time (ms)' } } },
      },
    });

    // Error Rate Chart
    new Chart(document.getElementById('errorChart'), {
      type: 'bar',
      data: {
        labels: timeLabels.slice(0, sampledErrors.length),
        datasets: [{
          label: 'Error Rate (%)',
          data: sampledErrors,
          backgroundColor: 'rgba(244, 67, 54, 0.7)',
          borderColor: '#F44336',
          borderWidth: 1,
        }],
      },
      options: {
        responsive: true,
        plugins: { legend: { position: 'top' } },
        scales: { y: { beginAtZero: true, max: 100 } },
      },
    });

    // Active Threads Chart
    new Chart(document.getElementById('threadsChart'), {
      type: 'line',
      data: {
        labels: timeLabels.slice(0, sampledThreads.length),
        datasets: [{
          label: 'Active Threads',
          data: sampledThreads,
          borderColor: '#4CAF50',
          backgroundColor: 'rgba(76, 175, 80, 0.1)',
          fill: true,
          tension: 0.3,
        }],
      },
      options: {
        responsive: true,
        plugins: { legend: { position: 'top' } },
        scales: { y: { title: { display: true, text: 'Number of Threads' } } },
      },
    });

    // Throughput Chart
    new Chart(document.getElementById('throughputChart'), {
      type: 'line',
      data: {
        labels: timeLabels.slice(0, sampledThroughput.length),
        datasets: [{
          label: 'Throughput (req/s)',
          data: sampledThroughput,
          borderColor: '#9C27B0',
          backgroundColor: 'rgba(156, 39, 176, 0.1)',
          tension: 0.3,
        }],
      },
      options: {
        responsive: true,
        plugins: { legend: { position: 'top' } },
        scales: { y: { beginAtZero: true } },
      },
    });
  </script>
</body>
</html>`;

  return html;
};

/**
 * Main execution
 */
const generateGraphs = async () => {
  console.log("\n╔════════════════════════════════════════════════════════════╗");
  console.log("║          🎯 SPIKE TEST GRAPH GENERATOR STARTING           ║");
  console.log("╚════════════════════════════════════════════════════════════╝\n");

  const jtlFiles = fs
    .readdirSync(RESULTS_DIR)
    .filter(f => f.endsWith(".jtl") && !f.endsWith("-aggregate.jtl"));

  for (const jtlFile of jtlFiles) {
    const filePath = path.join(RESULTS_DIR, jtlFile);
    const testName = path.basename(jtlFile, ".jtl");

    console.log(`\n📈 Processing: ${testName}`);
    
    const samples = parseJTL(filePath);
    if (!samples) {
      console.log(`   ⚠️  Could not parse ${jtlFile}`);
      continue;
    }

    const timeSeries = generateTimeSeries(samples, 2000); // 2-second bins
    
    if (timeSeries.length === 0) {
      console.log(`   ⚠️  No time-series data generated`);
      continue;
    }

    // Generate ASCII graphs for terminal
    generateASCIIGraph(`${testName} - P50 Response Time (ms)`, timeSeries, "p50", 80, 15);
    generateASCIIGraph(`${testName} - P95 Response Time (ms)`, timeSeries, "p95", 80, 15);
    generateASCIIGraph(`${testName} - Error Rate (%)`, timeSeries, "errorRate", 80, 10);
    generateASCIIGraph(`${testName} - Active Threads`, timeSeries, "activeThreads", 80, 15);

    // Generate HTML chart
    const htmlChart = generateHTMLChart(testName, timeSeries);
    const htmlPath = path.join(RESULTS_DIR, `${testName}-chart.html`);
    fs.writeFileSync(htmlPath, htmlChart);
    console.log(`   ✅ HTML chart: ${htmlPath}`);

    // Summary stats
    const allResponseTimes = samples.map(s => s.elapsed);
    const successCount = samples.filter(s => s.success).length;
    const errorRate = ((samples.length - successCount) / samples.length * 100).toFixed(2);

    console.log(`\n   📊 Summary Statistics:`);
    console.log(`      Total Samples: ${samples.length}`);
    console.log(`      Success Rate: ${(successCount / samples.length * 100).toFixed(2)}%`);
    console.log(`      Error Rate: ${errorRate}%`);
    console.log(`      P50 Response: ${percentile(allResponseTimes, 50)}ms`);
    console.log(`      P95 Response: ${percentile(allResponseTimes, 95)}ms`);
    console.log(`      Min Response: ${Math.min(...allResponseTimes)}ms`);
    console.log(`      Max Response: ${Math.max(...allResponseTimes)}ms`);
  }

  console.log(
    `\n✨ Graphs generated! Open HTML files in tests/spike/results/ to view interactive charts.\n`
  );
};

generateGraphs().catch(console.error);

File: tests/spike/graphs.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Spike Test Graph Generator (Python)
Generates professional performance graphs with dual Y-axes
"""

import csv
import os
import sys
from pathlib import Path
from datetime import datetime

# Force UTF-8 output on Windows
if sys.platform == 'win32':
    import io
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
from collections import defaultdict

try:
    import matplotlib.pyplot as plt
    import matplotlib.dates as mdates
    import numpy as np
except ImportError:
    print("[WARNING] matplotlib not installed. Run: pip install matplotlib numpy")
    exit(1)

# Configuration
RESULTS_DIR = Path(__file__).parent / "results"
OUTPUT_DIR = RESULTS_DIR
BIN_SIZE_MS = 2000  # 2-second bins


def parse_jml(jml_file):
    """Parse JTL CSV file"""
    samples = []
    try:
        with open(jml_file, 'r') as f:
            reader = csv.DictReader(f)
            for row in reader:
                try:
                    samples.append({
                        'timestamp': int(row.get('timeStamp', 0) or 0),
                        'elapsed': int(row.get('elapsed', 0) or 0),
                        'success': row.get('success', 'false').lower() == 'true',
                        'activeThreads': int(row.get('grpThreads', 0) or 0),
                    })
                except (ValueError, KeyError):
                    continue
    except Exception as e:
        print(f"   ❌ Error parsing {jml_file}: {e}")
        return []
    
    return samples


def bin_data(samples, bin_size_ms=2000):
    """Bin samples by time intervals"""
    if not samples:
        return []
    
    min_time = min(s['timestamp'] for s in samples)
    max_time = max(s['timestamp'] for s in samples)
    
    bins = []
    t = min_time
    while t <= max_time:
        bin_end = t + bin_size_ms
        bin_samples = [s for s in samples if t <= s['timestamp'] < bin_end]
        
        if bin_samples:
            response_times = [s['elapsed'] for s in bin_samples]
            success_count = sum(1 for s in bin_samples if s['success'])
            error_count = len(bin_samples) - success_count
            
            bins.append({
                'time': t,
                'time_str': f"{(t - min_time) / 1000:.1f}s",
                'p50': np.percentile(response_times, 50),
                'p95': np.percentile(response_times, 95),
                'avg': np.mean(response_times),
                'min': np.min(response_times),
                'max': np.max(response_times),
                'error_rate': (error_count / len(bin_samples)) * 100,
                'active_threads': bin_samples[0]['activeThreads'],
                'throughput': len(bin_samples),
                'sample_count': len(bin_samples),
            })
        
        t += bin_size_ms
    
    return bins


def generate_response_time_chart(test_name, samples):
    """Generate response time chart with dual Y-axes"""
    bins = bin_data(samples, BIN_SIZE_MS)
    if not bins:
        print(f"   [WARNING] No data to chart")
        return
    
    times = list(range(len(bins)))
    p50_values = [b['p50'] for b in bins]
    p95_values = [b['p95'] for b in bins]
    active_threads = [b['active_threads'] for b in bins]
    
    fig, ax1 = plt.subplots(figsize=(14, 6))
    
    # Left Y-axis: Response Time
    color1 = '#2196F3'
    ax1.set_xlabel('Time (seconds)', fontsize=11)
    ax1.set_ylabel('Response Time (ms)', color=color1, fontsize=11)
    line1 = ax1.plot(times, p50_values, color=color1, label='P50 Response Time', linewidth=2)
    line2 = ax1.plot(times, p95_values, color='#FF9800', label='P95 Response Time', linewidth=2, linestyle='--')
    ax1.tick_params(axis='y', labelcolor=color1)
    ax1.set_ylim(bottom=0)
    ax1.grid(True, alpha=0.3)
    
    # Right Y-axis: Active Threads
    color2 = '#4CAF50'
    ax2 = ax1.twinx()
    ax2.set_ylabel('Active Threads', color=color2, fontsize=11)
    line3 = ax2.plot(times, active_threads, color=color2, label='Active Threads', linewidth=2.5, linestyle=':')
    ax2.tick_params(axis='y', labelcolor=color2)
    ax2.set_ylim(bottom=0)
    
    # Legend
    lines = line1 + line2 + line3
    labels = [l.get_label() for l in lines]
    ax1.legend(lines, labels, loc='upper left', fontsize=10)
    
    plt.title(f'Response Time & Thread Load - {test_name}', fontsize=13, fontweight='bold', pad=20)
    plt.tight_layout()
    
    output_file = OUTPUT_DIR / f'{test_name}-response-time.png'
    plt.savefig(output_file, dpi=150, bbox_inches='tight')
    print(f"   [OK] {output_file.name}")
    plt.close()


def generate_error_rate_chart(test_name, samples):
    """Generate error rate chart with dual Y-axes"""
    bins = bin_data(samples, BIN_SIZE_MS)
    if not bins:
        return
    
    times = list(range(len(bins)))
    error_rates = [b['error_rate'] for b in bins]
    throughput = [b['throughput'] for b in bins]
    
    fig, ax1 = plt.subplots(figsize=(14, 6))
    
    # Left Y-axis: Throughput
    color1 = '#9C27B0'
    ax1.set_xlabel('Time (seconds)', fontsize=11)
    ax1.set_ylabel('Throughput (req/s)', color=color1, fontsize=11)
    line1 = ax1.bar(times, throughput, color=color1, alpha=0.7, label='Throughput')
    ax1.tick_params(axis='y', labelcolor=color1)
    ax1.set_ylim(bottom=0)
    ax1.grid(True, alpha=0.3, axis='y')
    
    # Right Y-axis: Error Rate
    color2 = '#F44336'
    ax2 = ax1.twinx()
    ax2.set_ylabel('Error Rate (%)', color=color2, fontsize=11)
    line2 = ax2.plot(times, error_rates, color=color2, label='Error Rate', linewidth=2.5, marker='o', markersize=4)
    ax2.tick_params(axis='y', labelcolor=color2)
    ax2.set_ylim(0, max(100, max(error_rates) * 1.2))
    
    # Legend
    lines = [line2[0]]
    labels = ['Error Rate']
    ax1.legend(lines, labels, loc='upper left', fontsize=10)
    
    plt.title(f'Error Rate & Throughput - {test_name}', fontsize=13, fontweight='bold', pad=20)
    plt.tight_layout()
    
    output_file = OUTPUT_DIR / f'{test_name}-error-rate.png'
    plt.savefig(output_file, dpi=150, bbox_inches='tight')
    print(f"   [OK] {output_file.name}")
    plt.close()


def generate_summary_table(test_name, samples):
    """Generate summary statistics table"""
    if not samples:
        return
    
    bins = bin_data(samples, 5000)  # Larger bins for table
    
    response_times = [s['elapsed'] for s in samples]
    success_count = sum(1 for s in samples if s['success'])
    error_count = len(samples) - success_count
    
    print(f"\n   {test_name.upper()} SUMMARY")
    print(f"   {'-' * 70}")
    print(f"   Total Samples:        {len(samples):,}")
    print(f"   Success:              {success_count:,} ({(success_count/len(samples)*100):.1f}%)")
    print(f"   Errors:               {error_count:,} ({(error_count/len(samples)*100):.1f}%)")
    print(f"   P50 Response Time:    {np.percentile(response_times, 50):.0f}ms")
    print(f"   P95 Response Time:    {np.percentile(response_times, 95):.0f}ms")
    print(f"   Min Response Time:    {min(response_times):.0f}ms")
    print(f"   Max Response Time:    {max(response_times):.0f}ms")
    print(f"   Peak Active Threads:  {max(s['activeThreads'] for s in samples)}")
    print(f"   {'-' * 70}")


def main():
    print("\n" + "="*60)
    print("  SPIKE TEST GRAPH GENERATOR (PYTHON)")
    print("="*60 + "\n")
    
    if not RESULTS_DIR.exists():
        print(f"ERROR: Results directory not found: {RESULTS_DIR}")
        return
    
    # Find JTL files
    jml_files = sorted(RESULTS_DIR.glob("*.jtl"))
    jml_files = [f for f in jml_files if not f.name.endswith("-aggregate.jtl")]
    
    if not jml_files:
        print(f"WARNING: No JTL files found in {RESULTS_DIR}")
        return
    
    print(f"Processing {len(jml_files)} test(s) from {RESULTS_DIR}\n")
    
    for jml_file in jml_files:
        test_name = jml_file.stem
        print(f"[*] {test_name}")
        
        samples = parse_jml(jml_file)
        if not samples:
            print(f"    WARNING: No valid samples")
            continue
        
        generate_response_time_chart(test_name, samples)
        generate_error_rate_chart(test_name, samples)
        generate_summary_table(test_name, samples)
    
    print(f"\nDone! Graphs generated in: {OUTPUT_DIR}")
    print(f"[GRAPHS] Open PNG files to view charts\n")


if __name__ == '__main__':
    main()
